### DISCUSSION TITLE
Scrollbehaviour weird jumping

### DISCUSSION THREAD

**Discussion by Gyurmatag** [nuxt/nuxt#17720] [Questions]
So I configured my rouse.scrollBehaviour.ts file in Nuxt 3. Here it is:

```
import { defineNuxtPlugin } from "#app";

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
    if (savedPosition) {
      return savedPosition;
    }
    return { left: 0, top: 0, behaviour: "smooth" };
  };
});
```

The problem is that I think it's scrolling earlier than the pages load, so there is a weird jumping effect on the pages.

Here is a video of the problem:
https://user-images.githubusercontent.com/22452368/154849559-3974fc01-e265-486b-865b-55ee03053fa8.mp4

Can you please help me what is the problem here? Or is there a bug in Nuxt 3?[](url)

**Comment by olivervorasai** [comment]
Would love to hear a solution for this as well.

**Comment by florian-lefebvre** [comment]
Inspired by https://github.com/zernonia/madewithsupabase, I do the following in my `app.vue` instead:

```vue
<script setup lang="ts">
const route = useRoute();
const firstVisit = ref(true);

watch(
  route,
  (n) => {
    if (process.client) {
      if (firstVisit.value) {
        firstVisit.value = false;
        return;
      }
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  },
  { deep: true, immediate: true }
);
</script>
```

**Reply by madsh93** [reply]
This is in my opinion still the best experience. I hope this issue will be resolved somewhat soon.

**Comment by madsh93** [comment]
As far as I am aware this is something planned to be fixed with:
https://github.com/nuxt/framework/pull/3485

 Is this correct @danielroe ?

**Reply by danielroe** [reply]
Exactly 👍

**Reply by madsh93** [reply]
```js
  router: {
    options: {
      scrollBehavior: function (to, from, savedPosition) {
        return { x: 0, y: 0 }
      }
    }
  },
  ```
  
  Might be wrong on the syntax, but shouldn't this be possible now?

**Reply by danielroe** [reply]
No. It's not serializable because it's a function. See https://v3.nuxtjs.org/docs/directory-structure/pages#router-options to see how to do it.

**Comment by florian-lefebvre** [comment]
The solution is to create a new file `app/router.options.ts` with the following content:
```ts
import type { RouterOptions } from "vue-router";

// https://router.vuejs.org/api/#routeroptions
export default <Partial<RouterOptions>>{
  scrollBehavior: (to, from, savedPosition) => {
    if (savedPosition) {
      return savedPosition;
    }
    return { top: 0 };
  },
};
```
Would be worth putting this in the docs.

**Reply by arhubi** [reply]
Thank's Florian for this suggestion. 
But actually, this tip doesn't seem to work well on Safari or Firefox. Scroll behavior is still weird as it is looks to be triggered too early (like if it was on the `from` page). It looks like it only works on Chromium browsers.

The recent merged PR about router configuration (#3485) didn't improve that.

I still haven't be able to figure out why it's happening, but it's quite annoying considering a smooth navigation to a new page or between the history.

**Reply by florian-lefebvre** [reply]
Is it actually a `nuxt3` or a `vue-router` issue?

**Reply by florian-lefebvre** [reply]
Right configuration is now:
```ts
import type { RouterConfig } from "@nuxt/schema";

// https://router.vuejs.org/api/#routeroptions
export default <RouterConfig>{
  scrollBehavior: (to, from, savedPosition) => {
    if (savedPosition) {
      return savedPosition;
    }
    if (to.hash) {
      return {
        el: to.hash,
      };
    }
    return { top: 0 };
  },
};
```

**Comment by joel-wenzel** [comment]
In nuxt 2 the ```triggerScroll``` event was emitted when the page was ready to be scrolled, which ultimately resolved the scrollBehavior promise. I am aware the event bus is gone in vue 3 but is there a similar hook we can use to resolve the scrollBehavior promise?

**Comment by joel-wenzel** [comment]
@danielroe I updated NuxtPage to emit a new hook `page:transition:finish` that works much better inside router.options `scrollBehavior`. 
```typescript
         _wrapIf(
              Transition,
              {
                ...(routeProps.route.meta.pageTransition ??
                  defaultPageTransition),
                onAfterLeave: () =>
                  nuxtApp.callHook(
                    'page:transition:finish',
                    routeProps.Component
                  ),
              },
```

relevant `app/router.options.ts`
```typescript
export default <RouterOptions>{
  scrollBehavior(to, from, savedPosition) {
    const nuxtApp = useNuxtApp()
   
    return new Promise(async (resolve) => {
      nuxtApp.hooks.hookOnce('page:transition:finish', async () => {
        await nextTick()
        resolve(position)
   }
}
```

I can submit a PR if you guys would be interested in this. It works quite well.

**Reply by danielroe** [reply]
Sure! ❤️ It seems like we need a better default scrollBehavior so maybe a PR that sets a better default scrollbehaviour + the hook?

**Reply by joel-wenzel** [reply]
created: https://github.com/nuxt/framework/pull/3851

**Reply by Gyurmatag** [reply]
When this will be merged? @danielroe
